morango/syncsession.py

```diff
@@ -0,0 +1,407 @@
+                                           buffer.model_name, buffer.profile, buffer.partition, buffer.conflicting_serialized_data, 1
+                                    FROM {buffer} AS buffer
+                                    WHERE buffer.transfer_session_id = '{transfer_session_id}'
+                           """.format(buffer=Buffer._meta.db_table,
+                                      store=Store._meta.db_table,
+                                      transfer_session_id=self.transfer_session_id)
+
+        cursor.execute(insert_remaining_buffer)
+
+    def _dequeuing_insert_remaining_rmcb(self, cursor):
+        # insert remaining records into rmc
+        insert_remaining_rmcb = """REPLACE INTO {rmc} (instance_id, counter, store_model_id)
+                                    SELECT rmcb.instance_id, rmcb.counter, rmcb.model_uuid
+                                    FROM {rmcb} AS rmcb
+                                    WHERE rmcb.transfer_session_id = '{transfer_session_id}'
+                           """.format(rmc=RecordMaxCounter._meta.db_table,
+                                      rmcb=RecordMaxCounterBuffer._meta.db_table,
+                                      transfer_session_id=self.transfer_session_id)
+
+        cursor.execute(insert_remaining_rmcb)
+
+    def _dequeuing_delete_remaining_rmcb(self, cursor):
+        # delete the rest for this transfer session
+        delete_remaining_rmcb = """DELETE FROM {rmcb}
+                                  WHERE {rmcb}.transfer_session_id = '{transfer_session_id}'
+                               """.format(rmcb=RecordMaxCounterBuffer._meta.db_table,
+                                          transfer_session_id=self.transfer_session_id)
+
+        cursor.execute(delete_remaining_rmcb)
+
+    def _dequeuing_delete_remaining_buffer(self, cursor):
+        delete_remaining_buffer = """DELETE FROM {buffer}
+                                 WHERE {buffer}.transfer_session_id = '{transfer_session_id}'
+                              """.format(buffer=Buffer._meta.db_table,
+                                         transfer_session_id=self.transfer_session_id)
+
+        cursor.execute(delete_remaining_buffer)
+
+    @transaction.atomic
+    def _integrate_into_store(self):
+        """
+        Takes data from the buffers and merges into the store and record max counters.
+        """
+        with connection.cursor() as cursor:
+            self._dequeuing_delete_rmcb_records(cursor)
+            self._dequeuing_delete_buffered_records(cursor)
+            self._dequeuing_merge_conflict_rmcb(cursor)
+            current_id = InstanceIDModel.get_and_update_current_instance()
+            self._dequeuing_merge_conflict_buffer(cursor, current_id)
+            self._dequeuing_update_rmcs_last_saved_by(cursor, current_id)
+            self._dequeuing_delete_mc_rmcb(cursor)
+            self._dequeuing_delete_mc_buffer(cursor)
+            self._dequeuing_insert_remaining_buffer(cursor)
```

Comment:
After deleting the merge conflicting buffers, all we have is fast forwards left and records were never in the store before. So I forego doing the fast forward queries and just insert the remaining records into the store.